### ISSUE TITLE
[otbn] Indirection in BN.LID / BN.SID

### ISSUE THREAD

**Issue by rswarbrick** [lowRISC/opentitan#3587]
If OTBN executes:
```
  bn.lid x10++, 0(x11)
```
and `x10 = 0x101`, what should happen?

I assume that we'll probably load data into `w1` (since `0x101 & 0x1f = 0x1`). We should probably document that. But what will be the new value of `x10`? (Is it `0x102` or `0x2`?)

> estimate 4

**Comment by felixmiller** [comment]
I'd say let's treat it as a 32 bit value and just take the lowest 5 bit for the reg reference. Just for the sake of reducing complexity. So my feeling is that we should load data in `w1` and next value should be `0x102`.

**Comment by GregAC** [comment]
I'd agree with @felixmiller no wrap around (beyond the usual wrap around when you hit the top of the 32-bit range) just use the bottom 5 bits and ignore the rest.

**Comment by rswarbrick** [comment]
Ok, cool. The reason I asked is because it means we need a 32-bit counter, rather than a 5-bit counter. But presumably we can re-use part of the ALU?

**Comment by felixmiller** [comment]
I also was thinking that the ALU is used, when I replied above. If this is a burden for HW, or a timing issue, the 5 bit boundary is also fine. But this would anyways only cover the pointer in the reg file, right? There is also the dmem pointer which I think would be 12 bit and which has to be incremented by 32.

**Comment by GregAC** [comment]
That is a good point, we have two adds here, one for the address calc, one for the increment so we will need an extra adder/incrementer.

OTBN is comparatively large so a 32-bit incrementer shouldn't have meaningful area impact though to keep it small we could say in addition to only considering the bottom 5 bits for getting a register number the ++ only operates on the bottom 5 bits also.

We could also use the stall cycle so we can reuse the main ALU for the increment, but this would mean adding a stall cycle to SID plus it's a slightly more complex execution of the instruction. I think we're better off with the extra area and less complex instruction execution.

**Comment by rswarbrick** [comment]
I think we've now got 3 plausible candidates:

1. 32-bit add
1. 5-bit add; zero extend
1. 5-bit add; leave other 27 bits unchanged ("MIMD!")

I don't think there's any good reason to add a stall cycle. I'm happy with any of the three but, if we can't re-use an existing adder, I'd err towards option 2 or 3.

Let's pick one and document it. We can always change our minds later.

**Comment by GregAC** [comment]
I think option 2. would be the minimal area option hardware wise, @felixmiller any preference from a SW perspective?

**Comment by felixmiller** [comment]
I think SW is agnostic to this, but it is not fully clear to me.
If we have an instruction of the type `bn.lid x10, 0(x11++)` the 5-bit counter could not be used for the increment. Or would this be another HW anyways? (This is independent from the address calculation with the offset, which is needed on-top)

**Comment by GregAC** [comment]
> There is also the dmem pointer which I think would be 12 bit and which has to be incremented by 32.

Ah yes had forgotten about this one, again we could define this in terms of an increment (ignore bits [4:0] increment bits [11:5], unaligned addresses where bits [4:0] !=0 already cause an error)

We can only do one increment so we could use a 7-bit incrementer (well set by DMem size increment but 7 bit for our current 4k dmem) with some muxing to select out the thing to increment. Again with the option of zeroing out the bits we don't touch in the increment or leave them as is (I think I'd tend towards zero out as it does simplify things a bit more if we're handling both styles of increment with one incrementer).

**Comment by imphil** [comment]
Regarding wrapping of DMEM addresses: It feels dangerous/fragile to have ISA/software behavior which depends on the DMEM size. Imagine a scenario where we have written a lot of software for OTBN, and then increase the DMEM size. This will cause existing software which relies on the wrapping behavior to behave differently (read different data).

I can see two ways of ISA behavior which don't depend on the size of DMEM. Are there others?
* Go with a full 32b adder (32b wrap-arounds) and only use the lower bits. (Don't we have enough adders in the design that we could re-use for the address calculation?)
* Or error out/issue an alert on an overflow (this will still require thinking about what to do with those unused bits).

**Comment by GregAC** [comment]
> Don't we have enough adders in the design that we could re-use for the address calculation?

Not really, we have one 32-bit adder on the base side, yes we have 2 256-bit ones on the wide side but I'd really rather not add further stuff in the data paths for them especially as it'll introduce a narrow side RF -> wide side ALU data path. These will be the timing critical paths of the design so we want to keep them as straight-forward as possible.

However the incrementer will be less area than the adder (each bit is a half-adder rather than full-adder as you only have two inputs current number and carry-in, rather than three, operand a & b and carry-in) so doing a full 32-bit increment isn't a big deal area wise (especially relative to overall OTBN area).

Though I think I'd be in favour of alert on overflow. 'No reasonable software will rely on wrapping behaviour for address increments' feels like an OK statement to make.

**Comment by rswarbrick** [comment]
> Though I think I'd be in favour of an error on overflow. 'No reasonable software will rely on wrapping behaviour for address increments' feels like an OK statement to make.

I'm fine with that. I think we should keep the wrapping for the register side, because that sounds like you might actually do it on purpose :-)

**Comment by rswarbrick** [comment]
Re-opening, because the PR doesn't sort out the "how should we do address wrapping" discussion.

**Comment by felixmiller** [comment]
I think this is really sth. to decide in favor for the HW. From the SW side for both dmem access and regfile access, SW won't rely on the wrapping. The indirect reg access, will always be only a subset of the WDRs being used to store a bigint (currently max is 16 WDRs for RSA 4096) and some other WDRs storing other quantities. So having a design parameter that does the wrapping at log2(DMEM_SIZE) and log2(REGFILE_SIZE) would not be a problem for SW.
I see @imphil's point though, that you want to avoid dependencies in the spec from the dmem size. To a certain extent this also holds for the regfile size. I don't think that someone would increase regfile size (we also wouldn't have the encoding space anyways) but reducing to 16 WDRs might be an option we want to evaluate at some point if a specific implementation does not need to support RSA 4096 (or only with reduced perf. requirements)

**Comment by GregAC** [comment]
So the current state of affairs in the RTL is:

* Register increments wrap at register address boundary (i.e. 31 -> 0)
* Memory increments wrap at dmem sized boundary

Neither trigger an error, but error handling is still to be implemented.

From what @felixmiller has just said SW won't be relying on wrapping anywhere so I'd be minded to say an error on either of the overflows that currently silently wrap.

**Comment by imphil** [comment]
My high-level view on it: 
* Wrapping behavior should not depend on parameters, i.e. the size of DMEM/IMEM. In other words, increasing the size of DMEM should not suddenly load different data from memory (when it wrapped before, but now doesn't).
* We don't need to and should not try to "recover" from what is likely to be a programmer's mistake. We can safely assume the developer has made a mistake when addressing out-of-range memory, and error out. 
  (This is in line with what [Felix said above](https://github.com/lowRISC/opentitan/issues/3587#issuecomment-702677463): "SW won't rely on the wrapping".)
* When writing "error out" I mean a recoverable alert.

Applying this to our ISA:
* For incrementable operands referencing a WDR
  * Applies to `<grd>[<grd_inc>]` in `BN.LID`, `<grs2>[<grs2_inc>]` in `BN.SID`, `<grd>[<grd_inc>]` and `<grs>[<grs_inc>]` in `BN.MOVR`
  * Currently, we document e.g. `grd is updated to be (*grd + 1) & 0x1f`, which would both do a 5 bit wrap-around and discard the upper bits which might have been present before in the GPR.
  * My preferred behavior:
	  * Alert on an overflow.
	  * Don't touch bits `GPR[n][31:5]` from a user's perspective:
        * If we have a tiny bit of RTL area to spare, error out if `GPR[n][31:5] != 0` to make it very clear that the GPR should only contain a reference to a WDR in the range 0..31, and everything else is considered a programming error.
	    * Otherwise, write back only the lower 5 bit, retain the higher bits.
* For incrementable operands referencing a data memory (DMEM) address:
  * Applies to `<offset>(<grs1>[<grs1_inc>])` in `BN.LID`, and `<offset>(<grs1>[<grs1_inc>])` in `BN.SID` (`<offset>` is 15 bit signed, i.e. in the range of -16384..16352)
  * Fun complication: `grs1` doesn't need to be a valid DMEM address itself (i.e. in the range 0..DMEM_SIZE-1), only `grs1 + offset` must be.
  * Currently, our documentation isn't overly specific, but I'd read it as 32b adder with 32b wrap-arounds: `GPR[rs1] = GPR[rs1] + (WLEN / 8)`. We implement, as [Greg said above](https://github.com/lowRISC/opentitan/issues/3587#issuecomment-726808667), a wrap-around at the DMEM size on `grs1` alone; `offset` is not taken into account, as noted in https://github.com/lowRISC/opentitan/pull/4727#issuecomment-754613936)
  * Implementation options:
    1. Do what the docs (intend to) say today: increment grs1 with a 32b adder and wrap-around (normal unsigned arithmetic). No alerts during the increment step. Check for valid addresses only before executing the BN.LID/BN.SID, then taking the offset into account.
    2. Require the base address to be in the DMEM size range (ignoring offset), and error out when trying to increment the address above this range. This is what Greg proposed in https://github.com/lowRISC/opentitan/pull/4727#issuecomment-754613936.
    3. Wrap around at the maximum value of grs1 that could lead to a valid DMEM address when also taking offset into account, which would be a wrap around at 15 bit for DMEM sizes < 16 kiB, as discussed at https://github.com/lowRISC/opentitan/pull/4727. Potentially error out instead of wrapping when exceeding the 2^15-1.
  * My preferred behavior: I'm not sure. 
    * Option (a) has the highest hardware overhead, but is easiest to understand. 
    * Option (b) could be a good middle ground and would match the behavior I'd prefer for operands referencing a WDR (described above), *if* software developers are OK with the fact that the base address (grs1 in BN.LID/BN.SID) must always be within 0..DMEM_SIZE-1. I'd like to hear from @felixmiller on that.
    * Option (c) seems rather odd and hard to describe to a software developer, so I'm least fond of it.

**Comment by rswarbrick** [comment]
For the DMEM operands, I don't think anyone is suggesting (c). The code in #4727 is supposed to implement (a): the only time that it doesn't wrap properly is when we'd fail because of a bad address, so we can compute any old rubbish: it will be ignored because we'll squash the write enables anyway.

**Comment by rswarbrick** [comment]
For the WDR case, good point about the upper bits. I think I'd probably prefer not to spot the "error" when wrapping from 31. Otherwise, you can't do things like loop up to `w31` where the body contains an increment. (The last iteration would wrap past 31, triggering an error).

**Comment by imphil** [comment]
> For the DMEM operands, I don't think anyone is suggesting (c). The code in #4727 is supposed to implement (a): the only time that it doesn't wrap properly is when we'd fail because of a bad address, so we can compute any old rubbish:

Doesn't https://github.com/lowRISC/opentitan/pull/4727 wrap at 15 bit? Yes, this implementation behaves  like a 32b adder as long as there is a sequence of BN.LID/BN.SID instructions and the n+1th instruction checks for a valid address before starting execution. However, we could also have a BN.LID followed by e.g. just a simple ADD using the incremented address in the GPR. In that case, the user sees that we did wrapping at 15 bit (i.e. "the maximum for DMEM_SIZE + offset"). (Or am I totally missing something in #4727? It's a confusing topic.)

> I think I'd probably prefer not to spot the "error" when wrapping at 32. Otherwise, you can't do things like loop up to w31 where the body contains an increment. 

Good and interesting use case. Before we find a solution to that: @felixmiller does that code look like something you'd write?

(Beyond that: @rswarbrick `bn.lid x2++, 0(x3++)` is invalid with two increments; only one is allowed.)

**Comment by rswarbrick** [comment]
> Doesn't #4727 wrap at 15 bit?

Barring bugs, that code is supposed to act like a full 32b adder, so long as the base + offset is a valid dmem address. If the address is not valid, the "next instruction" won't happen (because we'll have stopped with an error). If the address is valid, then the value the code computes *is* the same thing as you'd get with a full 32b adder.

> `bn.lid x2++, 0(x3++)` is invalid with two increments; only one is allowed.

Duh! But I guess you get the idea - something like this to load up striped data, maybe?
```
    addi x2, 16
    addi x3, 24
    addi x4, 0x100
    loopi 8, 3
      bn.lid x2++, 0(x4)
      bn.lid x3++, 32(x4)
      addi x4, x4, 64
```

**Comment by imphil** [comment]
> Barring bugs, that code is supposed to act like a full 32b adder, so long as the base + offset is a valid dmem address. If the address is not valid, the "next instruction" won't happen (because we'll have stopped with an error). If the address is valid, then the value the code computes is the same thing as you'd get with a full 32b adder.

Argh, of course ("Did I ever say anything else?"). I edited the summary in https://github.com/lowRISC/opentitan/issues/3587#issuecomment-756871416 to reflect that; sorry for the confusion.

Let's quickly pick this up in the meeting today, but I think we have agreement to go with option (a) for the case "DMEM operands" case (which is also what the spec says today, so no spec changes needed).

Let's also discuss the register case there, it looks like we're not quite there yet deciding between wrapping and alerting, and what should happen with the upper bits. (For everyone following along at home, I'll update this issue with what we discussed.)

**Comment by felixmiller** [comment]
I am happy with approach a) and it seems the most straight forward to me.
Looking at the `bn.sid` and `bn.lid` type of instruction, iirc the reason for having the offset in the first place was either because we said we have the ALU available anyways or to make it look more RISC-V-ish or both. I am not sure anymore tbh.

Looking at this purely from the SW perspective we never use the full capabilities of the instructions.

Currently we only have two variants how we use them:
1. `bn.lid x[0..31], <imm>(x0) `for loading a full WLEN word into a WREG from DMEM with hard coded address
2. `bn.lid x[0..31], 0(x[1..31])` for using a narrow reg as pointer into dmem

plus:
- `bn.sid `we only use in the 2nd variant currently.
- the 1st variant for `bn.lid` is only used in the transition phase of the code in the code snippets, we currently use it to load constants from IMEM via a DMEM detour. These constant will eventually end up in DMEM.

So from my perspective any of the following would also be fine:
- completely get rid of the offset
- requiring either the offset being zero or the address being zero (x0).
- allowing only predefined values for the offset (this might be helpful, if we added say a DROM or partition DMEM in any other way)

**Comment by rswarbrick** [comment]
Ok, cool. Let's go with "option (a)" for now and get the fixes in.

I'll file a follow-up issue about simplifying BN.LID/BN.SID where we can talk about whether we want to cut them down at all. Probably the big question is whether it's useful to do so for RTL implementation. I don't see any particular spec advantages, and the big wart (being able to encode increments on both counters) is there either way.

**Comment by rswarbrick** [comment]
Right, I'm closing this issue now. We have a spec and an implementation of that spec. We could probably improve things further, and I've split out questions about how to do that into separate issues (tracked above).

**Comment by GregAC** [comment]
So for clarity is option A the following?

> Do what the docs (intend to) say today: increment grs1 with a 32b adder and wrap-around (normal unsigned arithmetic). No alerts during the increment step. Check for valid addresses only before executing the BN.LID/BN.SID, then taking the offset into account. (#4727 is implementing that in an optimized way.)

**Comment by rswarbrick** [comment]
Yep, that's what I meant.

**Comment by felixmiller** [comment]
re: incrementing WDR references: @rswarbrick is making a good point with the [example](https://github.com/lowRISC/opentitan/issues/3587#issuecomment-756888621) code where the last loop iteration is wrapping over the 5 bit boundary. This wouldn't be an issue in current code since we are still using `w31` as all-zero reg, but I am planning to change this, to align it with `x0` purpose. Then it could become an issue and it would be weird if handling a slice up to `w31` needed special treatment.

I think if we really wanted to catch the an error here had to use a 6-bit WDR reference and then raise an alert whenever a GPR is incremented where any of the upper bits (31..5) is set. But this would complicate the spec, so having said this I am happy to move forward with the current spec.

**Comment by cdgori** [comment]
Two comments:
1) Seems like the re-alignment of zero-reg will make this wrapping behavior more likely/common?  So allowing the wrap around is necessary/useful.  I had originally thought we could just narrow the functionality to disallow this wrap but seems like that is not going to be the case.
2) I do think having any bits 31..5 set in the WDR reference is very bad and we should detect it.  I'm also not convinced that alert is sufficient here, maybe some more drastic action is needed, as we would have for detected glitches/faults.  (I am not generally sympathetic to "programmer error" here as this code should be very thoroughly tested, so things that should not occur should be flagged as such, because they are often indicators of / correlated to attack scenarios.)

**Comment by rswarbrick** [comment]
For anyone following along, I've copied Chris's last comment across to the relevant child issues. If we want to discuss things further, let's do it there.
